src/osd/OSD.cc

```diff
@@ -8881,18 +8889,35 @@ void OSD::do_recovery(
   ThreadPool::TPHandle &handle)
 {
   uint64_t started = 0;
-  if (cct->_conf->osd_recovery_sleep > 0) {
-    handle.suspend_tp_timeout();
-    pg->unlock();
-    utime_t t;
-    t.set_from_double(cct->_conf->osd_recovery_sleep);
-    t.sleep();
-    dout(20) << __func__ << " slept for " << t << dendl;
-    pg->lock();
-    handle.reset_tp_timeout();
+  if (cct->_conf->osd_recovery_sleep > 0 && service.recovery_needs_sleep) {
+    auto recovery_requeue_callback = new FunctionContext([this, pg](int r) {
+      pg->lock();
+      dout(20) << "do_recovery wake up at "
+               << ceph_clock_now()
+	       << ", re-queuing recovery" << dendl;
+      service.recovery_needs_sleep = false;
+      pg->recovery_queued = false;
+      pg->queue_recovery();
```

Comment:
At this point this PGQueuable already went through and was accounted for in recovery_ops_reserved when it was originally enqueued, but pg->queue_recovery() will end up calling OSDService::queue_for_recovery()
->OSDService::_maybe_queue_recovery() which will try to account it in recovery_ops_reserved again.

It seems like we can avoid taking the pg lock and changing any pg state here, and re-add the pg directly via OSDService::_queue_for_recovery() under the recovery_lock.